README.md
# vitality

- Welcome to Vitality, a dynamic Flutter package designed to effortlessly infuse your app with captivating live animations in the background. With just 10 lines of code, you can breathe life into your UI by adding randomly moving icons, circles, rectangles, images, and more. Let's explore how you can enhance your app's visual appeal with ease.

## Examples:

Check out these examples showcasing the versatility of the Vitality library:

- ![example 1](https://github.com/AbdulrhmanSayedAli/vitality/raw/main/example/images/example_1.gif) ![example 6](https://github.com/AbdulrhmanSayedAli/vitality/raw/main/example/images/example_6.gif)

- ![example 3](https://github.com/AbdulrhmanSayedAli/vitality/raw/main/example/images/example_3.gif) ![example 4](https://github.com/AbdulrhmanSayedAli/vitality/raw/main/example/images/example_4.gif)

- ![example 5](https://github.com/AbdulrhmanSayedAli/vitality/raw/main/example/images/example_5.gif) ![example 2](https://github.com/AbdulrhmanSayedAli/vitality/raw/main/example/images/example_2.gif)

## Usage:

- The Vitality package is designed with simplicity in mind, making integration a breeze. Take a look at the example below to quickly grasp how to use it:

```dart
Vitality.randomly(
  background: Colors.black,
  maxOpacity: 0.8,
  minOpacity: 0.3,
  itemsCount: 80,
  enableXMovements: false,
  whenOutOfScreenMode: WhenOutOfScreenMode.Teleport,
  maxSpeed: 1.5,
  maxSize: 30,
  minSpeed: 0.5,
  randomItemsColors: [Colors.yellowAccent, Colors.white],
  randomItemsBehaviours: [
    ItemBehaviour(shape: ShapeType.Icon, icon: Icons.star),
    ItemBehaviour(shape: ShapeType.Icon, icon: Icons.star_border),
    ItemBehaviour(shape: ShapeType.StrokeCircle),
  ],
)
```

## Parameters:

- `whenOutOfScreenMode` Determines how a shape behaves at the screen edge, values are:

  - `none`: It does nothing and continues its movements out the screen.
  - `Reflect`: It bounces and returns in the oppisite direction.
  - `Teleport`: It continues its movements to the other side of the screen.

- `randomItemsBehaviours` Defines the shapes that the library can generate. Choose from the `ShapeType` enum (`FilledCircle`, `StrokeCicle`, `FilledSquare`, `Icon`, `Image`, etc.) :

  - For `Icon` type, pass an `IconData` object:

    ```dart
     ItemBehaviour(shape: ShapeType.Icon, icon: Icons.star)
    ```

  - For `Image` type, pass a `dart:ui Image` object:

    ```dart
     ItemBehaviour(shape: ShapeType.Image, image: /*ui.Image() here*/)
    ```

    Sample code to load an image from assets:

    - Convert the path into `dart:ui Image` object

    ```dart
    import 'dart:ui' as ui;
    import 'package:flutter/services.dart';

    Future<ui.Image> loadImage(String assetPath) async {
      ByteData data = await rootBundle.load(assetPath);
      List<int> bytes = data.buffer.asUint8List();
      ui.Codec codec = await ui.instantiateImageCodec(Uint8List.fromList(bytes));
      ui.FrameInfo fi = await codec.getNextFrame();
      return fi.image;
    }
    ```

    - Integrate this function into your code :

    ```dart
    FutureBuilder<ui.Image>(
      future: loadImage('assets/images/your_image.png'),
      builder: (context, snapshot) {
        if (snapshot.connectionState == ConnectionState.done) {
          return Vitality.randomly(
            whenOutOfScreenMode: WhenOutOfScreenMode.Reflect,
            randomItemsColors: const [Colors.yellow],
            randomItemsBehaviours: [
              ItemBehaviour(
                shape: ShapeType.Image,
                image: snapshot.data!,
              )
            ],
          );
        }
        return const CircularProgressIndicator();
      },
    ),
    ```

## Full Customization:

- If you desire even more customization in your animations, Vitality provides a custom mode that is straightforward to use. In this mode, you can specify the exact shapes, icons, or images you want to animate using the `shapes` parameter. Additionally, you can control the background with the `background` parameter.

### Customizing Individual Shapes:

- To customize individual shapes, you can pass a list of `Shape` objects as the `shapes` parameter. Each `Shape` object includes the following properties:
  - `pos`: Initial position of the shape [`Offset`].
  - `dx`: Speed on the X-axis [`double`].
  - `dy`: Speed on the Y-axis [`double`].
  - `size`: Size of the shape [`double`].
  - `color`: Color of the shape [`Color`].
  - `whenOutOfScreenMode`: Behavior of the shape at the screen edge [`WhenOutOfScreenMode`].
  - `behaviour`: Structure of the shape [`ItemBehaviour`].

```dart
Vitality.custom(shapes: [
  Shape(
    pos: const Offset(30, 30),
    dx: 0.3,
    dy: 0,
    size: 160,
    color: Colors.yellow,
    whenOutOfScreenMode: WhenOutOfScreenMode.Reflect,
    behaviour: ItemBehaviour(shape: ShapeType.Icon, icon: Icons.sunny),
  ),
]),
```

### Customizing Groups of Shapes:

- For more dynamic combinations, you can use the `ShapesGenerator` class. This class allows you to generate shapes with specific properties,
  resulting in an easier way to seamlessly combine multiple shapes, allowing you to effortlessly craft a beautiful background. The parameters include:

  - `minWidth`, `minHeight`: Minimum width and height for random generation [double].
  - `maxWidth`, `maxHeigh`: Maximum width and height for random generation [double].
  - `enableXMovements`, `enableYMovements`: Whether or not the shapes will move on the X and Y axes [bool].
  - `minSize`, `maxSize`: Minimum and maximum size values for shapes [double].
  - `minOpacity`, `maxOpacity`: Minimum and maximum opacity values for shapes (values between 0 and 1) [double].
  - `minSpeed`, `maxSpeed`: Minimum and maximum speed values for shapes [double].
  - `behaviours`: List of `ItemBehaviour` for random selection.
  - `colors`: List of `Color` for random selection.
  - `whenOutOfScreenMode`: What will happen when shapes reach the screen edge [`WhenOutOfScreenMode`].

```dart
List<Shape> getClouds(Size size) {
  return ShapesGenerator.randomly(
    maxWidth: size.width,
    maxHeight: size.height / 3,
    enableYMovements: false,
    maxSize: 60,
    minSize: 50,
    maxOpacity: 0.9,
    minOpacity: 0.5,
    whenOutOfScreenMode: WhenOutOfScreenMode.Reflect,
    behaviours: [ItemBehaviour(shape: ShapeType.Icon, icon: Icons.cloud)],
    colors: const [Colors.white],
  ).getShapes(10);
}
```

Now, let's mix different shapes to make something truly magical.

```dart
List<Shape> getClouds(Size size) {
  return ShapesGenerator.randomly(
    maxWidth: size.width,
    maxHeight: size.height / 3,
    enableYMovements: false,
    maxSize: 60,
    minSize: 50,
    maxOpacity: 0.9,
    minOpacity: 0.5,
    whenOutOfScreenMode: WhenOutOfScreenMode.Reflect,
    behaviours: [ItemBehaviour(shape: ShapeType.Icon, icon: Icons.cloud)],
    colors: const [Colors.white],
  ).getShapes(10);
}

List<Shape> getGroundShapes(Size size) {
  return ShapesGenerator.randomly(
    minHeight: size.height - 130,
    maxWidth: size.width,
    maxHeight: size.height,
    enableYMovements: false,
    maxSize: 190,
    minSize: 90,
    maxOpacity: 1,
    minOpacity: 0.5,
    whenOutOfScreenMode: WhenOutOfScreenMode.none,
    behaviours: [
      ItemBehaviour(shape: ShapeType.Icon, icon: Icons.circle)
    ],
    colors: const [Colors.green],
    maxSpeed: 0,
    minSpeed: 0
  ).getShapes(40);
}



@override
Widget build(BuildContext context) {
  Size size = MediaQuery.of(context).size;
  return Vitality.custom(background: Colors.blue, shapes: [
    Shape(
      pos: const Offset(30, 30),
      dx: 0.3,
      dy: 0,
      size: 160,
      color: Colors.yellow,
      whenOutOfScreenMode: WhenOutOfScreenMode.Reflect,
      behaviour: ItemBehaviour(shape: ShapeType.Icon, icon: Icons.sunny),
    ),
    ...getClouds(size),
    ...getGroundShapes(size),
  ]);
}
```


CHANGELOG.md
# Changelog

All notable changes to this package will be documented in this file.

## 1.0.1

- A Flutter package that Allows you to create a very beautiful live animations in the background (like randomly moving icons ,circles ,rectangles ....) within 10 lines.

- added the randomly generated animations with a custom settings chosen by you (like make the movements vertically only, determine minimum and maximum speed, size, opacity ...).

- added the lines generated animation which gives animated queues if shapes with number of lines determined by you.

## 1.0.2

- Fixing error : "Unsupported operation: Cannot modify an unmodifiable list" error when passing a const array of colors

## 1.0.3 [Retracted Version]

- Improving performance

## 1.0.4

- Fixing error : "App error: setState() called after dispose():VitalityState#eee0e(lifecycle state: defunct, not mounted)"

## 2.0.0

- Removing the required width and height properties (Vitality will now take the parent's width and height).
- Improving performance.
- Fixing the error: sometimes the animations freeze.

## 2.1.0

- Enhancing functionality by introducing the capability to draw not only icons and shapes but also images.

## 3.0.0

- Introduced a custom mode option to enhance specific customization capabilities.
- Implemented minimum height and width parameters in the `ShapesGenerator` for greater control.
- Most `ShapesGenerator` parameters is no longer required.
- Renamed all `ShapeType` names from triangle to square.
- Eliminate the need to import individual files, simply import the main library file.
- Documenting part of the code to make it clearer and easier to understand.


lib/models/ItemBehaviour.dart
import 'package:flutter/cupertino.dart';
import 'dart:ui' as ui;

/// Represents the behavior of the shapes generated by the library, along with all the necessary data defining their characteristics.
///
/// When creating an instance of this class, provide the following parameters:
///
/// - [shape] An object of type [ShapeType] that specifies the structure of the shape.
/// - [icon] If [shape] is set to [ShapeType.Icon], you must pass an [IconData] object to this field.
/// - [image] If [shape] is set to [ShapeType.Image], you need to provide a [ui.Image] object to this field.
class ItemBehaviour {
  ShapeType shape;
  IconData? icon;
  ui.Image? image;

  ItemBehaviour({required this.shape, this.icon, this.image});
}

/// Enumerates various types of shapes that the library can handle.
///
/// Use this enum to specify the type of shape when working with the library's functionality.
enum ShapeType {
  FilledCircle,
  StrokeCircle,
  DoubleStrokeCircle,
  TripleStrokeCircle,
  FilledSquare,
  StrokeSquare,
  DoubleStrokeSquare,
  TripleStrokeSquare,
  Icon,
  Image,
}


lib/shapesManagement/Shape.dart
import 'package:equatable/equatable.dart';
import 'package:flutter/material.dart';
import 'package:vitality/models/ItemBehaviour.dart';
import 'package:vitality/models/WhenOutOfScreenMode.dart';

/// Represents the physical manifestation of a shape, such as a moving circle, icon, image, etc.
///
/// When creating an instance of this class, provide the following parameters:
///
/// - [pos] The initial position of the shape.
/// - [dx] The speed of the shape on the X-axis.
/// - [dy] The speed of the shape on the Y-axis.
/// - [size] The size of the shape.
/// - [color] The color of the shape.
/// - [whenOutOfScreenMode] The behavior of the shape at the screen edge (see [WhenOutOfScreenMode] for more details).
/// - [behaviour] The behavior of the shape at the screen edge (see [ItemBehaviour] for more details).
/// ignore: must_be_immutable
class Shape implements Equatable {
  Offset pos;
  double dx;
  double dy;
  double size;
  Color color;
  ItemBehaviour behaviour;
  WhenOutOfScreenMode whenOutOfScreenMode;

  Shape(
      {required this.pos,
      required this.dx,
      required this.dy,
      required this.size,
      required this.color,
      required this.whenOutOfScreenMode,
      required this.behaviour});

  /// Draws multiple circles with centers positioned at [pos].
  ///
  /// Parameters:
  /// - [canvas] The Canvas object from custom paint used for drawing the circles.
  /// - [sizes] List of radiuses for the circles to be drawn.
  void drawCircles(Canvas canvas, List<double> sizes) {
    for (double rad in sizes) canvas.drawCircle(pos, rad, getPaint());
  }

  /// Draws multiple squares with centers positioned at [pos].
  ///
  /// Parameters:
  /// - [canvas] The Canvas object from custom paint used for drawing the squares.
  /// - [sizes] List of sizes for the squares to be drawn.
  void drawSquares(Canvas canvas, List<double> sizes) {
    for (double d in sizes)
      canvas.drawRect(
          Rect.fromPoints(
              Offset(pos.dx - d, pos.dy - d), Offset(pos.dx + d, pos.dy + d)),
          getPaint());
  }

  void drawIcon(Canvas canvas) {
    if (behaviour.icon == null) return;
    TextPainter textPainter = TextPainter(textDirection: TextDirection.ltr);
    textPainter.text = TextSpan(
      text: String.fromCharCode(behaviour.icon!.codePoint),
      style: TextStyle(
        color: color,
        fontSize: this.size,
        fontFamily: behaviour.icon!.fontFamily,
        package: behaviour.icon!.fontPackage,
      ),
    );
    textPainter.layout();
    textPainter.paint(
        canvas, Offset(pos.dx - this.size / 2, pos.dy - this.size / 2));
  }

  void drawImage(Canvas canvas) {
    if (behaviour.image == null) return;
    double w = behaviour.image!.width.toDouble();
    double h = behaviour.image!.height.toDouble();
    double scaledW = this.size;
    double scaledH = (scaledW * h) / w;

    canvas.drawImageRect(
        behaviour.image!,
        Rect.fromLTWH(0, 0, w, h),
        Rect.fromPoints(Offset(pos.dx - scaledW / 2, pos.dy - scaledH / 2),
            Offset(pos.dx + scaledW / 2, pos.dy + scaledH / 2)),
        getPaint());
  }

  void draw(Canvas canvas) {
    switch (behaviour.shape) {
      case ShapeType.FilledCircle:
        drawCircles(canvas, [this.size / 2]);
        break;
      case ShapeType.StrokeCircle:
        drawCircles(canvas, [this.size / 2]);
        break;
      case ShapeType.DoubleStrokeCircle:
        drawCircles(canvas, [this.size / 2, this.size / 4]);
        break;
      case ShapeType.TripleStrokeCircle:
        drawCircles(canvas, [this.size / 2, this.size / 3, this.size / 6]);
        break;
      case ShapeType.FilledSquare:
        drawSquares(canvas, [this.size / 2]);
        break;
      case ShapeType.StrokeSquare:
        drawSquares(canvas, [this.size / 2]);
        break;
      case ShapeType.DoubleStrokeSquare:
        drawSquares(canvas, [this.size / 2, this.size / 4]);
        break;
      case ShapeType.TripleStrokeSquare:
        drawSquares(canvas, [this.size / 2, this.size / 3, this.size / 6]);
        break;
      case ShapeType.Icon:
        drawIcon(canvas);
        break;
      case ShapeType.Image:
        drawImage(canvas);
        break;
    }
  }

  Paint getPaint() {
    Paint paint = Paint();
    paint.color = color;
    if (behaviour.shape == ShapeType.FilledCircle ||
        behaviour.shape == ShapeType.FilledSquare)
      paint.style = PaintingStyle.fill;
    else
      paint.style = PaintingStyle.stroke;

    return paint;
  }

  void deltaNone(double width, double height) {
    double x = pos.dx + dx;
    double y = pos.dy + dy;
    pos = Offset(x, y);
  }

  void deltaTeleport(double width, double height) {
    double x = pos.dx + dx;
    double y = pos.dy + dy;
    pos = Offset(
        x < -size / 2
            ? width + size / 2
            : x > width + size / 2
                ? -size / 2
                : x,
        y < -size / 2
            ? height + size / 2
            : y > height + size / 2
                ? -size / 2
                : y);
  }

  void deltaReflect(double width, double height) {
    double x = pos.dx + dx;
    double y = pos.dy + dy;
    if (x < size / 2) {
      dx *= -1;
      x = size / 2;
    } else if (x > width - size / 2) {
      dx *= -1;
      x = width - size / 2;
    }
    if (y < size / 2) {
      dy *= -1;
      y = size / 2;
    } else if (y > height - size / 2) {
      dy *= -1;
      y = height - size / 2;
    }
    pos = Offset(x, y);
  }

  void delta(double width, double height) {
    if (whenOutOfScreenMode == WhenOutOfScreenMode.none)
      deltaNone(width, height);
    else if (whenOutOfScreenMode == WhenOutOfScreenMode.Teleport)
      deltaTeleport(width, height);
    else
      deltaReflect(width, height);
  }

  @override
  List<Object?> get props =>
      [pos, dx, dy, size, color, behaviour, whenOutOfScreenMode];

  @override
  bool? get stringify => true;
}


lib/shapesManagement/ShapesGenerator.dart
import 'dart:math';
import 'dart:ui';
import 'package:equatable/equatable.dart';
import 'package:flutter/material.dart';
import 'package:vitality/vitality.dart';

// ignore: must_be_immutable
class ShapesGenerator implements Equatable {
  double maxWidth;
  double maxHeight;
  double minWidth;
  double minHeight;
  double maxSize;
  double minSize;
  double maxOpacity;
  double minOpacity;
  double maxSpeed;
  double minSpeed;
  bool enableXMovements;
  bool enableYMovements;
  List<ItemBehaviour> behaviours = [];
  List<Color> colors = [];
  WhenOutOfScreenMode whenOutOfScreenMode;

  ShapesGenerator.randomly({
    required this.maxWidth,
    required this.maxHeight,
    this.minWidth = 0,
    this.minHeight = 0,
    this.enableXMovements = DefaultenableXMovements,
    this.enableYMovements = DefaultenableYMovements,
    this.maxSize = DefaultmaxSize,
    this.minSize = DefaultminSize,
    this.maxOpacity = DefaultmaxOpacity,
    this.minOpacity = DefaultminOpacity,
    required List<ItemBehaviour> behaviours,
    required List<Color> colors,
    this.maxSpeed = DefaultmaxSpeed,
    this.minSpeed = DefaultminSpeed,
    this.whenOutOfScreenMode = DefaultwhenOutOfScreenMode,
  }) {
    this.behaviours = [];
    this.colors = [];
    for (ItemBehaviour b in behaviours) {
      this.behaviours.add(b);
      this.behaviours.add(b);
    }
    for (Color c in colors) {
      this.colors.add(c);
      this.colors.add(c);
    }

    this.colors.shuffle();
    this.behaviours.shuffle();
  }

  Random rm = Random();

  double getDouble(double min, double max) =>
      rm.nextDouble() * (max - min) + min;

  double randomSpeed() {
    double speed = getDouble(minSpeed, maxSpeed);
    double direction = rm.nextInt(2) == 0 ? -1 : 1;
    return speed * direction;
  }

  List<Shape> getShapes(int count) {
    return List.generate(
        count,
        (index) => Shape(
            pos: Offset(
                getDouble(minWidth, maxWidth), getDouble(minHeight, maxHeight)),
            dx: enableXMovements ? randomSpeed() : 0,
            dy: enableYMovements ? randomSpeed() : 0,
            size: getDouble(minSize, maxSize),
            color: colors[getDouble(0, colors.length - 1).toInt()]
                .withOpacity(getDouble(minOpacity, maxOpacity)),
            behaviour: behaviours[getDouble(0, behaviours.length - 1).toInt()],
            whenOutOfScreenMode: whenOutOfScreenMode));
  }

  List<Shape> getLinesShapes(int lines) {
    List<Shape> res = [];
    double size = maxHeight / (2 * lines);

    double speed = randomSpeed();
    double space = ((maxWidth % size) + size) / ((maxWidth / size) - 1);
    double x = 0;
    ItemBehaviour behaviour =
        behaviours[getDouble(0, behaviours.length - 1).toInt()];

    Color color = colors[getDouble(0, colors.length - 1).toInt()]
        .withOpacity(getDouble(minOpacity, maxOpacity));

    while (x < maxWidth) {
      res.add(
        Shape(
            pos: Offset(x, size / 2),
            dx: speed,
            dy: 0,
            size: size,
            color: color,
            behaviour: behaviour,
            whenOutOfScreenMode: whenOutOfScreenMode),
      );
      x += size + space;
    }
    return res;
  }

  @override
  List<Object?> get props => [
        minWidth,
        minHeight,
        maxWidth,
        maxHeight,
        maxSize,
        minSize,
        maxOpacity,
        minOpacity,
        maxSpeed,
        minSpeed,
        enableXMovements,
        enableYMovements,
        behaviours,
        colors
      ];

  @override
  bool? get stringify => true;
}


lib/vitality.dart
library vitality;

export 'shapesManagement/Shape.dart';
export 'shapesManagement/ShapesGenerator.dart';
export 'models/WhenOutOfScreenMode.dart';
export 'models/ItemBehaviour.dart';

import 'dart:async';
import 'dart:math';
import 'package:flutter/material.dart';
import 'models/ItemBehaviour.dart';
import 'models/WhenOutOfScreenMode.dart';
import 'models/vitalityMode.dart';
import 'painting/Painter.dart';
import 'shapesManagement/Shape.dart';
import 'shapesManagement/ShapesGenerator.dart';

const int DefaultItemsCount = 60;
const int DefaultLines = 5;
const double DefaultmaxSize = 50;
const double DefaultminSize = 5;
const double DefaultmaxOpacity = 0.8;
const double DefaultminOpacity = 0.1;
const double DefaultmaxSpeed = 1;
const double DefaultminSpeed = 0;
const bool DefaultenableXMovements = true;
const bool DefaultenableYMovements = true;
const WhenOutOfScreenMode DefaultwhenOutOfScreenMode = WhenOutOfScreenMode.none;

// ignore: must_be_immutable
class Vitality extends StatefulWidget {
  int itemsCount = DefaultItemsCount;
  double maxOpacity = DefaultmaxOpacity;
  double minOpacity = DefaultminOpacity;
  double maxSize = DefaultmaxSize;
  double minSize = DefaultminSize;
  double maxSpeed = DefaultmaxSpeed;
  double minSpeed = DefaultminSpeed;
  bool enableXMovements = DefaultenableXMovements;
  bool enableYMovements = DefaultenableYMovements;
  WhenOutOfScreenMode whenOutOfScreenMode = DefaultwhenOutOfScreenMode;
  List<ItemBehaviour> randomItemsBehaviours = [];
  List<Color> randomItemsColors = [];
  int lines = DefaultLines;
  List<Shape> shapes = [];
  late VitalityMode mode;
  Color? background;

  Vitality.randomly(
      {this.itemsCount = DefaultItemsCount,
      this.maxSize = DefaultmaxSize,
      this.minSize = DefaultminSize,
      this.enableXMovements = DefaultenableXMovements,
      this.enableYMovements = DefaultenableYMovements,
      this.maxOpacity = DefaultmaxOpacity,
      this.minOpacity = DefaultminOpacity,
      this.maxSpeed = DefaultmaxSpeed,
      this.minSpeed = DefaultminSpeed,
      this.whenOutOfScreenMode = DefaultwhenOutOfScreenMode,
      required this.randomItemsBehaviours,
      required this.randomItemsColors,
      Key? key,
      this.background})
      : super(key: key) {
    mode = VitalityMode.Randomly;
  }

  Vitality.custom({required this.shapes, Key? key, this.background})
      : super(key: key) {
    mode = VitalityMode.Custom;
  }

  Vitality.lines(
      {Key? key,
      this.maxOpacity = DefaultmaxOpacity,
      this.minOpacity = DefaultminOpacity,
      this.maxSpeed = DefaultmaxSpeed,
      this.minSpeed = DefaultminSpeed,
      this.lines = DefaultLines,
      required this.randomItemsBehaviours,
      required this.randomItemsColors,
      this.background})
      : super(key: key) {
    mode = VitalityMode.Lines;
    whenOutOfScreenMode = WhenOutOfScreenMode.Teleport;
  }

  @override
  _VitalityState createState() => _VitalityState(
      maxSpeed: maxSpeed,
      minSpeed: minSpeed,
      count: itemsCount,
      enableXMovements: enableXMovements,
      enableYMovements: enableYMovements,
      maxSize: maxSize,
      minSize: minSize,
      randomItemsBehaviours: randomItemsBehaviours,
      whenOutOfScreenMode: whenOutOfScreenMode,
      mode: mode,
      lines: lines,
      randomItemsColors: randomItemsColors,
      maxOpacity: min(maxOpacity, 1),
      minOpacity: max(minOpacity, 0),
      shapes: shapes);
}

class _VitalityState extends State<Vitality> {
  double maxSize;
  double minSize;
  double maxOpacity;
  double minOpacity;
  int count;
  int lines;
  double maxSpeed;
  double minSpeed;
  WhenOutOfScreenMode whenOutOfScreenMode;
  List<ItemBehaviour> randomItemsBehaviours;
  List<Color> randomItemsColors;
  late List<Shape> shapes;
  bool enableXMovements;
  bool enableYMovements;
  VitalityMode mode;
  late ShapesGenerator generator;
  List<List<Shape>> linesShapes = [];
  bool finishedInitilization = false;
  late Timer timer;

  _VitalityState({
    required this.minSize,
    required this.whenOutOfScreenMode,
    required this.enableYMovements,
    required this.enableXMovements,
    required this.count,
    required this.lines,
    required this.randomItemsBehaviours,
    required this.randomItemsColors,
    required this.maxSize,
    required this.mode,
    required this.maxOpacity,
    required this.minOpacity,
    required this.minSpeed,
    required this.maxSpeed,
    this.shapes = const [],
  });

  void initShapes(double width, double height) {
    generator = ShapesGenerator.randomly(
        minHeight: 0,
        minWidth: 0,
        maxWidth: width,
        maxHeight: height,
        maxSize: maxSize,
        minSize: minSize,
        maxOpacity: maxOpacity,
        minOpacity: minOpacity,
        behaviours: randomItemsBehaviours,
        minSpeed: minSpeed,
        colors: randomItemsColors,
        enableXMovements: enableXMovements,
        enableYMovements: enableYMovements,
        maxSpeed: maxSpeed,
        whenOutOfScreenMode: whenOutOfScreenMode);

    if (shapes.isEmpty && mode == VitalityMode.Randomly)
      shapes = generator.getShapes(count);

    if (linesShapes.isEmpty && mode == VitalityMode.Lines)
      linesShapes =
          List.generate(lines, (index) => generator.getLinesShapes(lines));
  }

  void initTimer(double width, double height) {
    timer = Timer.periodic(Duration(milliseconds: 1000 ~/ 60), (t) {
      setState(() {
        shapes.forEach((element) {
          element.delta(width, height);
        });

        linesShapes.forEach(
          (s) {
            s.forEach(
              (element) {
                element.delta(width, height);
              },
            );
          },
        );
      });
    });
    finishedInitilization = true;
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (BuildContext context, BoxConstraints constraints) {
        double height = constraints.maxHeight;
        double width = constraints.maxWidth;

        if (!finishedInitilization) {
          initShapes(width, height);
          initTimer(width, height);
        }

        if (mode == VitalityMode.Randomly || mode == VitalityMode.Custom)
          return ClipRRect(
            child: CustomPaint(
              size: Size(width, height),
              painter: VitalityPainter(shapes, widget.background),
            ),
          );
        else
          return ClipRRect(
            child: Container(
              color: widget.background,
              child: Column(
                mainAxisAlignment: MainAxisAlignment.spaceAround,
                children: [
                  SizedBox(
                    width: width,
                    height: height / (2 * (lines + 1)),
                  ),
                  for (int i = 0; i < lines; i++) ...[
                    CustomPaint(
                      size: Size(width, height / (2 * (lines + 1))),
                      painter: VitalityPainter(linesShapes[i], null),
                    ),
                    SizedBox(
                      width: width,
                      height: height / (2 * (lines + 1)),
                    )
                  ]
                ],
              ),
            ),
          );
      },
    );
  }

  @override
  void dispose() {
    timer.cancel();
    super.dispose();
  }
}
